### PR TITLE
Usapt0 spin fix fcore

### PR DESCRIPTION
**Pull request by PhillCli**
## Description
This PR is a follow up of PR #1948 and resolves issue #1830.
It covers the corner case where zero spin block occurs in USAPT0 mp2_term due to freezing the core.

## Checklist
- [x] Tests added for any new features
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [ ] Ready for merge

### CODE REVIEW

File: psi4/src/psi4/libsapt_solver/usapt0.cc [hunk 1]
```diff
@@ -2600,15 +2657,15 @@ void USAPT0::mp2_terms() {
         if (na_ablock > 0) {
             dfh->fill_tensor("Aa_ar", Aa_ar, {aastart, aastart + na_ablock});
             dfh->fill_tensor("Da_ar", Da_ar, {aastart, aastart + na_ablock});
-            if (alpha_exchange_) dfh->fill_tensor("Ba_as", Ba_as, {aastart, aastart + na_ablock});
-            if (alpha_exchange_) dfh->fill_tensor("Ca_as", Ca_as, {aastart, aastart + na_ablock});
+            if (alpha_exchange_ && naa > 0 && nab > 0) dfh->fill_tensor("Ba_as", Ba_as, {aastart, aastart + na_ablock});
+            if (alpha_exchange_ && naa > 0 && nab > 0) dfh->fill_tensor("Ca_as", Ca_as, {aastart, aastart + na_ablock});
```
[the remark this answers is not in the document]
Comment (PhillCli) [reply psi4/src/psi4/libsapt_solver/usapt0.cc]: simplified the conditional, thanks for the review!

File: psi4/src/psi4/libsapt_solver/usapt0.cc [hunk 2]
```diff
@@ -2600,15 +2657,15 @@ void USAPT0::mp2_terms() {
...
         }
 
         if (nb_ablock > 0) {
             dfh->fill_tensor("Ab_ar", Ab_ar, {bastart, bastart + nb_ablock});
             dfh->fill_tensor("Db_ar", Db_ar, {bastart, bastart + nb_ablock});
-            if (beta_exchange_) dfh->fill_tensor("Bb_as", Bb_as, {bastart, bastart + nb_ablock});
-            if (beta_exchange_) dfh->fill_tensor("Cb_as", Cb_as, {bastart, bastart + nb_ablock});
+            if (beta_exchange_ && nba > 0 && nbb > 0) dfh->fill_tensor("Bb_as", Bb_as, {bastart, bastart + nb_ablock});
+            if (beta_exchange_ && nba > 0 && nbb > 0) dfh->fill_tensor("Cb_as", Cb_as, {bastart, bastart + nb_ablock});
```
[the remark this answers is not in the document]
Comment (PhillCli) [reply psi4/src/psi4/libsapt_solver/usapt0.cc]: changed according to the suggestion.

File: psi4/src/psi4/libsapt_solver/usapt0.cc [hunk 3]
```diff
@@ -2618,15 +2675,19 @@ void USAPT0::mp2_terms() {
             if (na_bblock > 0) {
                 dfh->fill_tensor("Aa_bs", Aa_bs, {abstart, abstart + na_bblock});
                 dfh->fill_tensor("Da_bs", Da_bs, {abstart, abstart + na_bblock});
-                if (alpha_exchange_) dfh->fill_tensor("Ba_br", Ba_br, {abstart, abstart + na_bblock});
-                if (alpha_exchange_) dfh->fill_tensor("Ca_br", Ca_br, {abstart, abstart + na_bblock});
+                if (alpha_exchange_ && nab > 0 && naa > 0)
+                    dfh->fill_tensor("Ba_br", Ba_br, {abstart, abstart + na_bblock});
+                if (alpha_exchange_ && nab > 0 && naa > 0)
+                    dfh->fill_tensor("Ca_br", Ca_br, {abstart, abstart + na_bblock});
```
[the remark this answers is not in the document]
Comment (PhillCli) [reply psi4/src/psi4/libsapt_solver/usapt0.cc]: changed according to the suggestion.

File: psi4/src/psi4/libsapt_solver/usapt0.cc [hunk 4]
```diff
@@ -2618,15 +2675,19 @@ void USAPT0::mp2_terms() {
...
 
             if (nb_bblock > 0) {
                 dfh->fill_tensor("Ab_bs", Ab_bs, {bbstart, bbstart + nb_bblock});
                 dfh->fill_tensor("Db_bs", Db_bs, {bbstart, bbstart + nb_bblock});
-                if (beta_exchange_) dfh->fill_tensor("Bb_br", Bb_br, {bbstart, bbstart + nb_bblock});
-                if (beta_exchange_) dfh->fill_tensor("Cb_br", Cb_br, {bbstart, bbstart + nb_bblock});
+                if (beta_exchange_ && nbb > 0 && nba > 0)
+                    dfh->fill_tensor("Bb_br", Bb_br, {bbstart, bbstart + nb_bblock});
+                if (beta_exchange_ && nbb > 0 && nba > 0)
+                    dfh->fill_tensor("Cb_br", Cb_br, {bbstart, bbstart + nb_bblock});
```
[the remark this answers is not in the document]
Comment (PhillCli) [reply psi4/src/psi4/libsapt_solver/usapt0.cc]: changed according to the suggestion.
